lib/astarte_data_access/xandra_utils.ex

```diff
@@ -23,67 +23,120 @@ defmodule Astarte.DataAccess.XandraUtils do
 
   require Logger
 
-  @spec run(String.t(), (Xandra.Connection.t(), String.t() -> any)) ::
-          any | {:error, :invalid_realm_name}
-  def run(realm, fun) do
-    with :ok <- verify_realm(realm) do
-      keyspace = CQLUtils.realm_name_to_keyspace_name(realm, Config.astarte_instance_id!())
+  @cluster_default_name :astarte_data_access_xandra
 
-      Xandra.Cluster.run(:astarte_data_access_xandra, &fun.(&1, keyspace))
+  def validate_realm_name(realm_name) do
+    if Realm.valid_name?(realm_name) do
+      :ok
+    else
+      Logger.warning("Invalid realm name.",
+        tag: "invalid_realm_name",
+        realm: realm_name
+      )
+
+      {:error, :invalid_realm_name}
+    end
+  end
+
+  def build_keyspace_name!(keyspace_name),
+    do: CQLUtils.realm_name_to_keyspace_name(keyspace_name, Config.astarte_instance_id!())
+
+  def run(realm_name, fun, opts \\ []) when is_function(fun) do
```

Comment:
Breaking: this function does not perform realm name validation anymore. For example, `Astarte.DataAccess.Device.interface_version/3` did check if the realm name was valid, but does so no more.

I would suggest to keep the standard behaviour of `run` as it was, and use `opts` to explicitly disallow validation (as needed in e.g. `realm.ex`). This change would also keep backwards compatibility. Contextually, `validated_run/3` would be removed.